{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "707610fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gradio as gr\n",
    "import os\n",
    "import pandas as pd\n",
    "from langchain.llms import OpenAI\n",
    "from langchain.prompts import PromptTemplate\n",
    "from langchain.chains import LLMChain\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain.tools import BaseTool, StructuredTool, Tool, tool\n",
    "\n",
    "from langchain.memory import ChatMessageHistory\n",
    "\n",
    "from langchain.agents import initialize_agent\n",
    "from langchain.agents import AgentType\n",
    "from langchain.agents import load_tools\n",
    "#from langchain.tools import PubmedQueryRun\n",
    "from langchain.utilities import WikipediaAPIWrapper\n",
    "from langchain.agents import create_pandas_dataframe_agent\n",
    "\n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import random\n",
    "\n",
    "#set API key\n",
    "os.environ[\"OPENAI_API_KEY\"] = \"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\" #openai\n",
    "#prep database of reactions\n",
    "biochem_data_url = \"https://bkms.brenda-enzymes.org/download/Reactions_BKMS.tar.gz\"\n",
    "biochem_df = pd.read_csv(biochem_data_url, compression='gzip', header=0, sep=\"\\t\")\n",
    "\n",
    "def biosynthesize(molecule,organism):\n",
    "    \n",
    "    molecule = 'taxadiene'\n",
    "    organism = \"e coli\"\n",
    "    \n",
    "    #initialize LLM\n",
    "    llm = OpenAI(temperature=0.5)\n",
    "\n",
    "    #initialize memory\n",
    "    history = ChatMessageHistory()\n",
    "\n",
    "    #create prompt template for an openAI search\n",
    "    prompt = PromptTemplate(\n",
    "        input_variables=[\"molecule\",],\n",
    "        template=\"how do you biosynthesize {molecule}. Output a table of necessary genes with two columns: Gene and Function\",\n",
    "    )\n",
    "    #and initiate its associated chain\n",
    "    chain = LLMChain(llm=llm, prompt=prompt)\n",
    "    this_output = (chain.run(molecule))\n",
    "\n",
    "    #parse this output into a dataframe\n",
    "    this_output = this_output.split('\\n')\n",
    "    df_array = []\n",
    "    for line in this_output:\n",
    "        if len(line) > 0:\n",
    "            df_dict = {}\n",
    "            line = line.split(' | ')\n",
    "            if line[0] != 'Gene' or line[0] != 'gene' or line[0] != '----': #fucking weird ass edge cases\n",
    "                df_dict['gene'] = line[0]\n",
    "                if len(line) > 1:\n",
    "                    df_dict['function'] = line[1]\n",
    "                df_array.append(df_dict)\n",
    "\n",
    "\n",
    "    #next augment this table by asking wikipedia\n",
    "    #initialize tools\n",
    "    #schema for wikipedia search\n",
    "    class SearchInput(BaseModel):\n",
    "        query: str = Field(description=\"should be molecule name\")\n",
    "\n",
    "    wikipedia = WikipediaAPIWrapper()\n",
    "    tools = [\n",
    "        Tool.from_function(\n",
    "            func=wikipedia.run,\n",
    "            name = \"SearchWikipedia\",\n",
    "            description=\"Searching wikipedia for genes that produce a molecule\",\n",
    "            args_schema=SearchInput\n",
    "            # coroutine= ... <- you can specify an async method if desired as well\n",
    "        ),\n",
    "    ]\n",
    "    #initialize agent for the wikipedia search \n",
    "    #agent = initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True)\n",
    "    #agent.run('how do you biosynthesize {}. output the gene names.'.format(molecule))\n",
    "    \n",
    "    gene_df = pd.DataFrame(df_array)\n",
    "\n",
    "    #pull out just the needed columns from the reaction database for now to make it easier\n",
    "    biochem_reactions = biochem_df[['Reaction', 'Reaction_ID_MetaCyc', 'Recommended_Name']]\n",
    "\n",
    "    #initialize the pandas agent \n",
    "    pandas_agent = create_pandas_dataframe_agent(OpenAI(temperature=0), biochem_reactions, verbose=False)\n",
    "\n",
    "    #figure out the EC number of the gene\n",
    "    new_array = []\n",
    "    for i, gene in gene_df.iterrows():\n",
    "        answer = pandas_agent.run('which element most closely matches {}'.format(gene['function']))\n",
    "\n",
    "        biochem_info = biochem_df[(biochem_df.Reaction == answer) | (biochem_df.Reaction_ID_MetaCyc == answer) | (biochem_df.Recommended_Name == answer)]\n",
    "        #see if anything was even found\n",
    "        if biochem_info.shape[0] > 0:\n",
    "            new_dict = {}\n",
    "            new_dict['EC_Number'] = biochem_info.iloc[0]['EC_Number']\n",
    "            new_dict['gene'] = gene['gene']\n",
    "            new_dict['function'] = gene['function']\n",
    "            new_array.append(new_dict)\n",
    "    EC_df = pd.DataFrame(new_array)\n",
    "    \n",
    "    #find protein sequences for these predictions\n",
    "    Entrez.email = 'jayman1466@gmail.com'\n",
    "\n",
    "    gb_array = []\n",
    "    for i,gene in EC_df.iterrows():\n",
    "    #search NCBI for any record available\n",
    "        gb_dict = {}\n",
    "        info = Entrez.einfo() \n",
    "        info = Entrez.esearch(db = \"protein\",term = gene['EC_Number'])\n",
    "        record = Entrez.read(info)\n",
    "\n",
    "        #arbitrarily pick the first ID for now\n",
    "        this_id = record['IdList'][0]\n",
    "\n",
    "        #pull the sequence of that gene   \n",
    "        handle = Entrez.efetch(\n",
    "        db = \"protein\", id = this_id, rettype = \"fasta\")\n",
    "        record = SeqIO.read( handle, \"fasta\" ) \n",
    "        gb_dict['sequence'] = str(record.seq)\n",
    "        gb_dict['actual_name'] = record.name\n",
    "        gb_dict['ID'] = record.id\n",
    "        gb_dict['EC'] = gene['EC_Number']\n",
    "        gb_dict['desired_name'] = gene['gene']\n",
    "        gb_array.append(gb_dict)\n",
    "\n",
    "        \n",
    "        \n",
    "    #start to create the actual genbank file\n",
    "        \n",
    "    #to create random DNA sequences\n",
    "    def DNA(length):\n",
    "        return ''.join(random.choice('CGTA') for _ in range(length))\n",
    "\n",
    "    #back translate\n",
    "    def orf_recode(sequence):\n",
    "        sequence = sequence.upper()\n",
    "\n",
    "        this_dict = {\"A\":\"GCA\",\"C\":\"TGT\",\"D\":\"GAT\",\"E\":\"GAA\",\"F\":\"TTT\",\"G\":\"GGA\",\"H\":\"CAC\",\"I\":\"ATA\",\"K\":\"AAA\",\"L\":\"CTT\",\"M\":\"ATG\",\"N\":\"AAT\",\"P\":\"CCC\",\"Q\":\"CAA\",\"R\":\"AGG\",\"S\":\"TCA\",\"T\":\"ACA\",\"V\":\"GTA\",\"Y\":\"TAT\",\"W\":\"TGG\",\"*\":\"TAA\"};  \n",
    "\n",
    "        output=''\n",
    "\n",
    "        for c in sequence:\n",
    "            output+=this_dict[c]        \n",
    "        output+='TAA'\n",
    "        return output\n",
    "\n",
    "\n",
    "    current_DNA_coord = 0\n",
    "    this_name = molecule\n",
    "    #initiate a record for a genbank file output\n",
    "    record = SeqRecord(Seq(\"\"),\n",
    "                       id='1243',\n",
    "                       name=this_name,\n",
    "                       description='Generated with synXNA',\n",
    "                       annotations={\"molecule_type\": \"DNA\"})\n",
    "    full_sequence = \"\"\n",
    "\n",
    "    #iterate through the individual genes\n",
    "    for item in gb_array:\n",
    "        #promoter\n",
    "        pro = DNA(50)\n",
    "        feature = SeqFeature(FeatureLocation(start=current_DNA_coord, end=current_DNA_coord + len(pro)), type='promoter', qualifiers={'label':'promoter'})\n",
    "        record.features.append(feature)\n",
    "        full_sequence = full_sequence + pro\n",
    "        current_DNA_coord = current_DNA_coord + len(pro)\n",
    "\n",
    "        rbs = DNA(35)\n",
    "        feature = SeqFeature(FeatureLocation(start=current_DNA_coord, end=current_DNA_coord + len(rbs)), type='RBS', qualifiers={'label':'RBS'})\n",
    "        record.features.append(feature)\n",
    "        full_sequence = full_sequence + rbs\n",
    "        current_DNA_coord = current_DNA_coord + len(rbs)\n",
    "\n",
    "        cds = item['sequence']\n",
    "        cds_nuc = orf_recode(cds)\n",
    "        feature = SeqFeature(FeatureLocation(start=current_DNA_coord, end=current_DNA_coord + len(cds_nuc)), type='CDS', qualifiers={'label':item['EC'] + \" \" + item['desired_name'], 'translation':cds})\n",
    "        record.features.append(feature)\n",
    "        full_sequence = full_sequence + cds_nuc\n",
    "        current_DNA_coord = current_DNA_coord + len(cds_nuc)\n",
    "\n",
    "    record.seq = Seq(full_sequence)\n",
    "\n",
    "    #write genbank file\n",
    "    #output_file = open('{} Biosynthesis.gb'.format(molecule), 'w')\n",
    "    #SeqIO.write(record, output_file, 'genbank')\n",
    "    #output_file.close()\n",
    "    return(record)\n",
    "\n",
    "#create the interface\n",
    "demo = gr.Interface(\n",
    "    biosynthesize,\n",
    "    inputs=[gr.Textbox(lines=1, placeholder=\"Biosynthesize this molecule\"),gr.Textbox(lines=1, placeholder=\"In this bacteria\")],\n",
    "    outputs=\"text\",\n",
    "    title=\"What do you want to make\",\n",
    ")\n",
    "demo.launch()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
