{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0edec0bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"OPENAI_API_KEY\"] = \"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\" #openai\n",
    "os.environ[\"API_TOKEN\"] = \"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\" #huggin face"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abb40a83",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jayma\\AppData\\Local\\Temp\\ipykernel_26080\\3554654682.py:3: DtypeWarning: Columns (0,3,4,5,6,7,8,9,10,11,12,13,14,15,17,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  biochem_data = pd.read_csv(biochem_data_url, compression='gzip', header=0, sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "biochem_data_url = \"https://bkms.brenda-enzymes.org/download/Reactions_BKMS.tar.gz\"\n",
    "biochem_data = pd.read_csv(biochem_data_url, compression='gzip', header=0, sep=\"\\t\")\n",
    "biochem_data.to_csv('metadata/bio_chem_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a791ac4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.llms import OpenAI\n",
    "from langchain.prompts import PromptTemplate\n",
    "from langchain.chains import LLMChain\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain.tools import BaseTool, StructuredTool, Tool, tool\n",
    "\n",
    "from langchain.memory import ChatMessageHistory\n",
    "\n",
    "from langchain.agents import initialize_agent\n",
    "from langchain.agents import AgentType\n",
    "from langchain.agents import load_tools\n",
    "#from langchain.tools import PubmedQueryRun\n",
    "from langchain.utilities import WikipediaAPIWrapper\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import json\n",
    "import requests\n",
    "\n",
    "molecule = 'taxadiene'\n",
    "organism = \"e coli\"\n",
    "\n",
    "#initialize LLM\n",
    "llm = OpenAI(temperature=0.5)\n",
    "\n",
    "#initialize memory\n",
    "history = ChatMessageHistory()\n",
    "\n",
    "#create prompt template for an openAI search\n",
    "prompt = PromptTemplate(\n",
    "    input_variables=[\"molecule\",],\n",
    "    template=\"how do you biosynthesize {molecule}. Output a table of necessary genes with two columns: Gene and Function\",\n",
    ")\n",
    "#and initiate its associated chain\n",
    "chain = LLMChain(llm=llm, prompt=prompt)\n",
    "this_output = (chain.run(molecule))\n",
    "\n",
    "#parse this output into a dataframe\n",
    "this_output = this_output.split('\\n')\n",
    "df_array = []\n",
    "for line in this_output:\n",
    "    if len(line) > 0:\n",
    "        df_dict = {}\n",
    "        line = line.split(' | ')\n",
    "        if line[0] != 'Gene' or line[0] != 'gene' or line[0] != '----': #fucking weird ass edge cases\n",
    "            df_dict['gene'] = line[0]\n",
    "            if len(line) > 1:\n",
    "                df_dict['function'] = line[1]\n",
    "            df_array.append(df_dict)\n",
    "        \n",
    "\n",
    "#next augment this table by asking wikipedia\n",
    "#initialize tools\n",
    "#schema for wikipedia search\n",
    "class SearchInput(BaseModel):\n",
    "    query: str = Field(description=\"should be molecule name\")\n",
    "\n",
    "wikipedia = WikipediaAPIWrapper()\n",
    "tools = [\n",
    "    Tool.from_function(\n",
    "        func=wikipedia.run,\n",
    "        name = \"SearchWikipedia\",\n",
    "        description=\"Searching wikipedia for genes that produce a molecule\",\n",
    "        args_schema=SearchInput\n",
    "        # coroutine= ... <- you can specify an async method if desired as well\n",
    "    ),\n",
    "]\n",
    "#initialize agent for the wikipedia search \n",
    "#agent = initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True)\n",
    "#agent.run('how do you biosynthesize {}. output the gene names.'.format(molecule))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "84a6a06d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jayma\\AppData\\Local\\Temp\\ipykernel_26080\\2054964236.py:6: DtypeWarning: Columns (1,4,5,6,7,8,9,10,11,12,13,14,15,16,18,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  biochem_df = pd.read_csv('metadata/bio_chem_data.csv', header=0, sep=\",\")\n"
     ]
    }
   ],
   "source": [
    "from langchain.agents import create_pandas_dataframe_agent\n",
    "\n",
    "gene_df = pd.DataFrame(df_array)\n",
    "\n",
    "#import the biochemical data\n",
    "biochem_df = pd.read_csv('metadata/bio_chem_data.csv', header=0, sep=\",\")\n",
    "#pull out just the needed column for now to make it easier\n",
    "biochem_reactions = biochem_df[['Reaction', 'Reaction_ID_MetaCyc', 'Recommended_Name']]\n",
    "\n",
    "#initialize the pandas agent \n",
    "pandas_agent = create_pandas_dataframe_agent(OpenAI(temperature=0), biochem_reactions, verbose=False)\n",
    "\n",
    "#figure out the EC number of the gene\n",
    "new_array = []\n",
    "for i, gene in gene_df.iterrows():\n",
    "    answer = pandas_agent.run('which element most closely matches {}'.format(gene['function']))\n",
    "    \n",
    "    biochem_info = biochem_df[(biochem_df.Reaction == answer) | (biochem_df.Reaction_ID_MetaCyc == answer) | (biochem_df.Recommended_Name == answer)]\n",
    "    #see if anything was even found\n",
    "    if biochem_info.shape[0] > 0:\n",
    "        new_dict = {}\n",
    "        new_dict['EC_Number'] = biochem_info.iloc[0]['EC_Number']\n",
    "        new_dict['gene'] = gene['gene']\n",
    "        new_dict['function'] = gene['function']\n",
    "        new_array.append(new_dict)\n",
    "EC_df = pd.DataFrame(new_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e625f93a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC_Number</th>\n",
       "      <th>gene</th>\n",
       "      <th>function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4.2.3.17</td>\n",
       "      <td>Taxadiene Synthase</td>\n",
       "      <td>Enzyme that catalyzes the formation of taxadie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.5.1.29</td>\n",
       "      <td>Geranyl Diphosphate Synthase</td>\n",
       "      <td>Enzyme that catalyzes the formation of geranyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.5.1.28</td>\n",
       "      <td>Isopentenyl Diphosphate Synthase</td>\n",
       "      <td>Enzyme that catalyzes the formation of isopent...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  EC_Number                              gene  \\\n",
       "0  4.2.3.17                Taxadiene Synthase   \n",
       "1  2.5.1.29      Geranyl Diphosphate Synthase   \n",
       "2  2.5.1.28  Isopentenyl Diphosphate Synthase   \n",
       "\n",
       "                                            function  \n",
       "0  Enzyme that catalyzes the formation of taxadie...  \n",
       "1  Enzyme that catalyzes the formation of geranyl...  \n",
       "2  Enzyme that catalyzes the formation of isopent...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EC_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "66447e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "#find the protein sequence from ncbi \n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "Entrez.email = 'jayman1466@gmail.com'\n",
    "\n",
    "gb_array = []\n",
    "for i,gene in EC_df.iterrows():\n",
    "#search NCBI for any record available\n",
    "    gb_dict = {}\n",
    "    info = Entrez.einfo() \n",
    "    info = Entrez.esearch(db = \"protein\",term = gene['EC_Number'])\n",
    "    record = Entrez.read(info)\n",
    "\n",
    "    #arbitrarily pick the first ID for now\n",
    "    this_id = record['IdList'][0]\n",
    "\n",
    "    #pull the sequence of that gene   \n",
    "    handle = Entrez.efetch(\n",
    "    db = \"protein\", id = this_id, rettype = \"fasta\")\n",
    "    record = SeqIO.read( handle, \"fasta\" ) \n",
    "    gb_dict['sequence'] = str(record.seq)\n",
    "    gb_dict['actual_name'] = record.name\n",
    "    gb_dict['ID'] = record.id\n",
    "    gb_dict['EC'] = gene['EC_Number']\n",
    "    gb_dict['desired_name'] = gene['gene']\n",
    "    gb_array.append(gb_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2da912ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       Biosynthesis            4725 bp    DNA              UNK 01-JAN-1980\n",
      "DEFINITION  Generated with synXNA.\n",
      "ACCESSION   1243\n",
      "VERSION     1243\n",
      "KEYWORDS    .\n",
      "SOURCE      .\n",
      "  ORGANISM  .\n",
      "            .\n",
      "FEATURES             Location/Qualifiers\n",
      "     promoter        1..50\n",
      "                     /label=\"promoter\"\n",
      "     RBS             51..85\n",
      "                     /label=\"RBS\"\n",
      "     CDS             86..2674\n",
      "                     /label=\"4.2.3.17 Taxadiene Synthase\"\n",
      "                     /translation=\"MAQLSFNAALKMNALGNKAIHDPTNCRAKSEGQMMWVCSKSGRTR\n",
      "                     VKMSRGSGGPGPVVMMSSSTGTSKVVSETSSTIVDDIPRLSANYHGDLWHHNVIQTLET\n",
      "                     PFRESSTYQERADELVVKIKDMFNALGDGDISPSAYDTAWVARVATISSDGSEKPRFPQ\n",
      "                     ALNWVFNNQLQDGSWGIESHFSLCDRLLNTTNSVIALSVWKTGHSQVEQGTEFIAENLR\n",
      "                     LLNEEDELSPDFEIIFPALLQKAKALGINLPYDLPFIKYLSTTREARLTDVSAAADNIP\n",
      "                     ANMLNALEGLEEVMDWKKIMRFQSKDGSFLSSPASTACVLMNTGDEKCFTFLNNLLVKF\n",
      "                     GGCVPCMYSIDLLERLSLVDNIEHLGIGRHFKQEIKVALDYVYRHWSERGIGWGRDSLV\n",
      "                     PDLNTTALGLRTLRTHGYDVSSDVLNNFKDENGRFFSSAGQTHVELRSVVILFRASDLA\n",
      "                     FPDEGAMDDARKFAEPYLRDALATKISTNTKLFKEIEYVVEYPWHMSIPRSEARSYIDS\n",
      "                     YDDDYVWERKTLYRMPSLSNSKCLELAKLDFNIVQSLHQEELKLLTRWWKESGMADINF\n",
      "                     TRHRVAEVYFSSATFEPEYSATRIAFTKIGCLQVLFDDMADIFATLDELKSFTEGVKRW\n",
      "                     DTSLLHEIPECMQTCFKVWFKLIEEVNNDVVKVQGRDMLAHIRKPWELYFNCYVQEREW\n",
      "                     LDAGYIPTFEEYLKTYAISVGLGPCTLQPILLMGELVKDDVVEKVHYPSNMFELVSLSW\n",
      "                     RLTNDTKTYQAEKARGQQASGIACYMKDNLGATEEDAIKHICRVVDRALKEASFEYFKP\n",
      "                     SNDIPMGCKSFIFNLRLCVQIFYKFIDGYGIANEEIKDYIRKVYIDPIQV\"\n",
      "     promoter        2675..2724\n",
      "                     /label=\"promoter\"\n",
      "     RBS             2725..2759\n",
      "                     /label=\"RBS\"\n",
      "     CDS             2760..3728\n",
      "                     /label=\"2.5.1.29 Geranyl Diphosphate Synthase\"\n",
      "                     /translation=\"MRVEEVAAATDRYLLRALRGSPASLYGAASHLVRHGGKRLRPYMV\n",
      "                     VRSCQMLGGTMRGAMPAAAAVEMVHNFTLVHDDMMDNDEVRHGVPTVHEKFGGDAAILA\n",
      "                     GDVLFSRAFGALASSRIGPDGVAGMVSRLAGACAEVCEGQQADIEMGRSGRIPSEAAYV\n",
      "                     AMISKKTAALFEVSCAMGAISASAPRRDVARLASFGRNLGIAFQMTDDLMGTVGDSGAT\n",
      "                     GKPVGNDIREGKKSLPVLLALGAARGAERRALRAALGNARATARQIGSAADVMRSLGVD\n",
      "                     AEVRRRARTYARRAGRSLAPYGGPARDDLARLLDRVVERVS\"\n",
      "     promoter        3729..3778\n",
      "                     /label=\"promoter\"\n",
      "     RBS             3779..3813\n",
      "                     /label=\"RBS\"\n",
      "     CDS             3814..4725\n",
      "                     /label=\"2.5.1.28 Isopentenyl Diphosphate Synthase\"\n",
      "                     /translation=\"MSSLVLQCWKLSSPSLILQQNTSISMGAFKGIHKLQIPNSPLTVS\n",
      "                     ARGLNKISCSLNLQTEKLCYEDNDNDLDEELMPKHIALIMDGNRRWAKDKGLEVYEGHK\n",
      "                     HIIPKLKEICDISSKLGIQIITAFAFSTENWKRSKEEVDFLLQMFEEIYDEFSRSGVRV\n",
      "                     SIIGCKSDLPMTLQKCIALTEETTKGNKGLHLVIALNYGGYYDILQATKSIVNKAMNGL\n",
      "                     LDVEDINKNLFDQELESKCPNPDLLIRTGGEQRVSNFLLWQLAYTEFYFTNTLFPDFGE\n",
      "                     EDLKEAIMNFQQRHRRFGGHTY\"\n",
      "ORIGIN\n",
      "        1 gacccgttta agagcccggg cgtccctcgt gtttagctac gttgctggaa gaaattccaa\n",
      "       61 cgcaagagca gcggaccctt tgaatatggc acaactttca tttaatgcag cacttaaaat\n",
      "      121 gaatgcactt ggaaataaag caatacacga tcccacaaat tgtagggcaa aatcagaagg\n",
      "      181 acaaatgatg tgggtatgtt caaaatcagg aaggacaagg gtaaaaatgt caaggggatc\n",
      "      241 aggaggaccc ggacccgtag taatgatgtc atcatcaaca ggaacatcaa aagtagtatc\n",
      "      301 agaaacatca tcaacaatag tagatgatat acccaggctt tcagcaaatt atcacggaga\n",
      "      361 tctttggcac cacaatgtaa tacaaacact tgaaacaccc tttagggaat catcaacata\n",
      "      421 tcaagaaagg gcagatgaac ttgtagtaaa aataaaagat atgtttaatg cacttggaga\n",
      "      481 tggagatata tcaccctcag catatgatac agcatgggta gcaagggtag caacaatatc\n",
      "      541 atcagatgga tcagaaaaac ccaggtttcc ccaagcactt aattgggtat ttaataatca\n",
      "      601 acttcaagat ggatcatggg gaatagaatc acacttttca ctttgtgata ggcttcttaa\n",
      "      661 tacaacaaat tcagtaatag cactttcagt atggaaaaca ggacactcac aagtagaaca\n",
      "      721 aggaacagaa tttatagcag aaaatcttag gcttcttaat gaagaagatg aactttcacc\n",
      "      781 cgattttgaa ataatatttc ccgcacttct tcaaaaagca aaagcacttg gaataaatct\n",
      "      841 tccctatgat cttcccttta taaaatatct ttcaacaaca agggaagcaa ggcttacaga\n",
      "      901 tgtatcagca gcagcagata atatacccgc aaatatgctt aatgcacttg aaggacttga\n",
      "      961 agaagtaatg gattggaaaa aaataatgag gtttcaatca aaagatggat catttctttc\n",
      "     1021 atcacccgca tcaacagcat gtgtacttat gaatacagga gatgaaaaat gttttacatt\n",
      "     1081 tcttaataat cttcttgtaa aatttggagg atgtgtaccc tgtatgtatt caatagatct\n",
      "     1141 tcttgaaagg ctttcacttg tagataatat agaacacctt ggaataggaa ggcactttaa\n",
      "     1201 acaagaaata aaagtagcac ttgattatgt atataggcac tggtcagaaa ggggaatagg\n",
      "     1261 atggggaagg gattcacttg tacccgatct taatacaaca gcacttggac ttaggacact\n",
      "     1321 taggacacac ggatatgatg tatcatcaga tgtacttaat aattttaaag atgaaaatgg\n",
      "     1381 aaggtttttt tcatcagcag gacaaacaca cgtagaactt aggtcagtag taatactttt\n",
      "     1441 tagggcatca gatcttgcat ttcccgatga aggagcaatg gatgatgcaa ggaaatttgc\n",
      "     1501 agaaccctat cttagggatg cacttgcaac aaaaatatca acaaatacaa aactttttaa\n",
      "     1561 agaaatagaa tatgtagtag aatatccctg gcacatgtca atacccaggt cagaagcaag\n",
      "     1621 gtcatatata gattcatatg atgatgatta tgtatgggaa aggaaaacac tttataggat\n",
      "     1681 gccctcactt tcaaattcaa aatgtcttga acttgcaaaa cttgatttta atatagtaca\n",
      "     1741 atcacttcac caagaagaac ttaaacttct tacaaggtgg tggaaagaat caggaatggc\n",
      "     1801 agatataaat tttacaaggc acagggtagc agaagtatat ttttcatcag caacatttga\n",
      "     1861 acccgaatat tcagcaacaa ggatagcatt tacaaaaata ggatgtcttc aagtactttt\n",
      "     1921 tgatgatatg gcagatatat ttgcaacact tgatgaactt aaatcattta cagaaggagt\n",
      "     1981 aaaaaggtgg gatacatcac ttcttcacga aatacccgaa tgtatgcaaa catgttttaa\n",
      "     2041 agtatggttt aaacttatag aagaagtaaa taatgatgta gtaaaagtac aaggaaggga\n",
      "     2101 tatgcttgca cacataagga aaccctggga actttatttt aattgttatg tacaagaaag\n",
      "     2161 ggaatggctt gatgcaggat atatacccac atttgaagaa tatcttaaaa catatgcaat\n",
      "     2221 atcagtagga cttggaccct gtacacttca acccatactt cttatgggag aacttgtaaa\n",
      "     2281 agatgatgta gtagaaaaag tacactatcc ctcaaatatg tttgaacttg tatcactttc\n",
      "     2341 atggaggctt acaaatgata caaaaacata tcaagcagaa aaagcaaggg gacaacaagc\n",
      "     2401 atcaggaata gcatgttata tgaaagataa tcttggagca acagaagaag atgcaataaa\n",
      "     2461 acacatatgt agggtagtag atagggcact taaagaagca tcatttgaat attttaaacc\n",
      "     2521 ctcaaatgat atacccatgg gatgtaaatc atttatattt aatcttaggc tttgtgtaca\n",
      "     2581 aatattttat aaatttatag atggatatgg aatagcaaat gaagaaataa aagattatat\n",
      "     2641 aaggaaagta tatatagatc ccatacaagt ataaatccgc atcacaggac tcttttccgt\n",
      "     2701 atggtaagcg cttcttttgt atcttctagc gtaccaattc atcgatgttt acaggataaa\n",
      "     2761 tgagggtaga agaagtagca gcagcaacag ataggtatct tcttagggca cttaggggat\n",
      "     2821 cacccgcatc actttatgga gcagcatcac accttgtaag gcacggagga aaaaggctta\n",
      "     2881 ggccctatat ggtagtaagg tcatgtcaaa tgcttggagg aacaatgagg ggagcaatgc\n",
      "     2941 ccgcagcagc agcagtagaa atggtacaca attttacact tgtacacgat gatatgatgg\n",
      "     3001 ataatgatga agtaaggcac ggagtaccca cagtacacga aaaatttgga ggagatgcag\n",
      "     3061 caatacttgc aggagatgta cttttttcaa gggcatttgg agcacttgca tcatcaagga\n",
      "     3121 taggacccga tggagtagca ggaatggtat caaggcttgc aggagcatgt gcagaagtat\n",
      "     3181 gtgaaggaca acaagcagat atagaaatgg gaaggtcagg aaggataccc tcagaagcag\n",
      "     3241 catatgtagc aatgatatca aaaaaaacag cagcactttt tgaagtatca tgtgcaatgg\n",
      "     3301 gagcaatatc agcatcagca cccaggaggg atgtagcaag gcttgcatca tttggaagga\n",
      "     3361 atcttggaat agcatttcaa atgacagatg atcttatggg aacagtagga gattcaggag\n",
      "     3421 caacaggaaa acccgtagga aatgatataa gggaaggaaa aaaatcactt cccgtacttc\n",
      "     3481 ttgcacttgg agcagcaagg ggagcagaaa ggagggcact tagggcagca cttggaaatg\n",
      "     3541 caagggcaac agcaaggcaa ataggatcag cagcagatgt aatgaggtca cttggagtag\n",
      "     3601 atgcagaagt aaggaggagg gcaaggacat atgcaaggag ggcaggaagg tcacttgcac\n",
      "     3661 cctatggagg acccgcaagg gatgatcttg caaggcttct tgatagggta gtagaaaggg\n",
      "     3721 tatcataata cctattacta acaaggcctt tctaagcaaa acgttcacca ttcctccctc\n",
      "     3781 tcccctacta gtacttcctc gtagccgcat atgatgtcat cacttgtact tcaatgttgg\n",
      "     3841 aaactttcat caccctcact tatacttcaa caaaatacat caatatcaat gggagcattt\n",
      "     3901 aaaggaatac acaaacttca aatacccaat tcacccctta cagtatcagc aaggggactt\n",
      "     3961 aataaaatat catgttcact taatcttcaa acagaaaaac tttgttatga agataatgat\n",
      "     4021 aatgatcttg atgaagaact tatgcccaaa cacatagcac ttataatgga tggaaatagg\n",
      "     4081 aggtgggcaa aagataaagg acttgaagta tatgaaggac acaaacacat aatacccaaa\n",
      "     4141 cttaaagaaa tatgtgatat atcatcaaaa cttggaatac aaataataac agcatttgca\n",
      "     4201 ttttcaacag aaaattggaa aaggtcaaaa gaagaagtag attttcttct tcaaatgttt\n",
      "     4261 gaagaaatat atgatgaatt ttcaaggtca ggagtaaggg tatcaataat aggatgtaaa\n",
      "     4321 tcagatcttc ccatgacact tcaaaaatgt atagcactta cagaagaaac aacaaaagga\n",
      "     4381 aataaaggac ttcaccttgt aatagcactt aattatggag gatattatga tatacttcaa\n",
      "     4441 gcaacaaaat caatagtaaa taaagcaatg aatggacttc ttgatgtaga agatataaat\n",
      "     4501 aaaaatcttt ttgatcaaga acttgaatca aaatgtccca atcccgatct tcttataagg\n",
      "     4561 acaggaggag aacaaagggt atcaaatttt cttctttggc aacttgcata tacagaattt\n",
      "     4621 tattttacaa atacactttt tcccgatttt ggagaagaag atcttaaaga agcaataatg\n",
      "     4681 aattttcaac aaaggcacag gaggtttgga ggacacacat attaa\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#create genbank file \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import random\n",
    "\n",
    "#create random DNA sequences\n",
    "def DNA(length):\n",
    "    return ''.join(random.choice('CGTA') for _ in range(length))\n",
    "\n",
    "#back translate\n",
    "def orf_recode(sequence):\n",
    "    sequence = sequence.upper()\n",
    "        \n",
    "    this_dict = {\"A\":\"GCA\",\"C\":\"TGT\",\"D\":\"GAT\",\"E\":\"GAA\",\"F\":\"TTT\",\"G\":\"GGA\",\"H\":\"CAC\",\"I\":\"ATA\",\"K\":\"AAA\",\"L\":\"CTT\",\"M\":\"ATG\",\"N\":\"AAT\",\"P\":\"CCC\",\"Q\":\"CAA\",\"R\":\"AGG\",\"S\":\"TCA\",\"T\":\"ACA\",\"V\":\"GTA\",\"Y\":\"TAT\",\"W\":\"TGG\",\"*\":\"TAA\"};  \n",
    "    \n",
    "    output=''\n",
    "\n",
    "    for c in sequence:\n",
    "        output+=this_dict[c]        \n",
    "    output+='TAA'\n",
    "    return output\n",
    "\n",
    "\n",
    "current_DNA_coord = 0\n",
    "this_name = 'Biosynthesis'\n",
    "#initiate a record for a genbank file output\n",
    "record = SeqRecord(Seq(\"\"),\n",
    "                   id='1243',\n",
    "                   name=this_name,\n",
    "                   description='Generated with synXNA',\n",
    "                   annotations={\"molecule_type\": \"DNA\"})\n",
    "full_sequence = \"\"\n",
    "\n",
    "#iterate through the individual genes\n",
    "for item in gb_array:\n",
    "    #promoter\n",
    "    pro = DNA(50)\n",
    "    feature = SeqFeature(FeatureLocation(start=current_DNA_coord, end=current_DNA_coord + len(pro)), type='promoter', qualifiers={'label':'promoter'})\n",
    "    record.features.append(feature)\n",
    "    full_sequence = full_sequence + pro\n",
    "    current_DNA_coord = current_DNA_coord + len(pro)\n",
    "    \n",
    "    rbs = DNA(35)\n",
    "    feature = SeqFeature(FeatureLocation(start=current_DNA_coord, end=current_DNA_coord + len(rbs)), type='RBS', qualifiers={'label':'RBS'})\n",
    "    record.features.append(feature)\n",
    "    full_sequence = full_sequence + rbs\n",
    "    current_DNA_coord = current_DNA_coord + len(rbs)\n",
    "    \n",
    "    cds = item['sequence']\n",
    "    cds_nuc = orf_recode(cds)\n",
    "    feature = SeqFeature(FeatureLocation(start=current_DNA_coord, end=current_DNA_coord + len(cds_nuc)), type='CDS', qualifiers={'label':item['EC'] + \" \" + item['desired_name'], 'translation':cds})\n",
    "    record.features.append(feature)\n",
    "    full_sequence = full_sequence + cds_nuc\n",
    "    current_DNA_coord = current_DNA_coord + len(cds_nuc)\n",
    "\n",
    "record.seq = Seq(full_sequence)\n",
    "\n",
    "#write genbank file\n",
    "output_file = open('{} Biosynthesis.gb'.format(molecule), 'w')\n",
    "SeqIO.write(record, output_file, 'genbank')\n",
    "output_file.close()\n",
    "\n",
    "f = open('{} Biosynthesis.gb'.format(molecule), \"r\")\n",
    "print(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "a0c83487",
   "metadata": {},
   "outputs": [],
   "source": [
    "#search NCBI for genome genbank file for the  requested organism\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"jayman1466@gmail.com\"\n",
    "handle = Entrez.einfo() # or esearch, efetch, ...\n",
    "record = Entrez.read(handle)\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "6c34dff1",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'mmseqs2'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[84], line 6\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m \n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mast\u001b[39;00m\n\u001b[1;32m----> 6\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmmseqs2\u001b[39;00m\n\u001b[0;32m      8\u001b[0m \u001b[38;5;66;03m#directory containing genbank files \u001b[39;00m\n\u001b[0;32m      9\u001b[0m input_dir \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenomes/\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'mmseqs2'"
     ]
    }
   ],
   "source": [
    "#pull high value UTRs from bacterial genomes \n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import pandas as pd \n",
    "import ast\n",
    "import mmseqs2\n",
    "\n",
    "#directory containing genbank files \n",
    "input_dir = 'genomes/'\n",
    "\n",
    "#cutoff for minimum intergenic region\n",
    "igr_cutoff = 100\n",
    "\n",
    "#create dict of genomes and phyla in database\n",
    "species_dict = {}\n",
    "\n",
    "#create temp housekeeping gene fasta\n",
    "housekeeping_df = pd.read_csv('metadata/housekeeping_genes.csv', sep = \",\", header=0)\n",
    "with open('temp/housekeeping.fasta', 'w') as f:\n",
    "    for i, gene in housekeeping_df.iterrows():\n",
    "        name = gene['gene']\n",
    "        sequence = gene['sequence']\n",
    "        f.write(f'>{name}\\n{sequence}\\n')\n",
    "\n",
    "#open the genbank files\n",
    "for filename in os.listdir(input_dir):\n",
    "    \n",
    "    #append to list of species being considered\n",
    "    species_name = filename[:filename.find('.')]\n",
    "    species_dict[filename] = species_name #fix this later\n",
    "        \n",
    "    gb_record = SeqIO.read(open(input_dir+filename,\"r\"), \"genbank\")\n",
    "    gb_sequence = gb_record.seq\n",
    "    \n",
    "    #create dataframe for these features\n",
    "    features_array = []\n",
    "    for feature in gb_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            while True:\n",
    "                try: \n",
    "                    features_dict = {}\n",
    "                    features_dict['name'] = feature.qualifiers['locus_tag'][0]\n",
    "                    features_dict['translation'] = str(feature.extract(gb_sequence).translate(table=11,cds=True))\n",
    "                    features_dict['start'] = int(feature.location.start)\n",
    "                    features_dict['end'] = int(feature.location.end)\n",
    "                    features_dict['strand'] = feature.location.strand\n",
    "                    features_array.append(features_dict)\n",
    "                    break\n",
    "                except Exception: \n",
    "                    break\n",
    "    features_df = pd.DataFrame(features_array)\n",
    "    features_df = features_df.sort_values('start')\n",
    "    \n",
    "    #look just for non-overlapping (in the same direction) genes\n",
    "    non_overlap_genes = []\n",
    "    \n",
    "    features_plus = features_df.query('strand == 1').reset_index(drop = True)    \n",
    "    for i,feature in features_plus.iterrows():\n",
    "        if i > 0:\n",
    "            previous_end = features_plus.iloc[i-1]['end']\n",
    "            this_start = feature['start']\n",
    "            if this_start - previous_end >= igr_cutoff:\n",
    "                non_overlap_genes.append(feature['name'])\n",
    "                \n",
    "    features_minus = features_df.query('strand == -1').reset_index(drop = True)\n",
    "    for i,feature in features_minus.iterrows():\n",
    "        if i < features_minus.shape[0]-1:\n",
    "            next_start = features_minus.iloc[i+1]['start']\n",
    "            this_end = feature['end']\n",
    "            if next_start - this_end >= igr_cutoff:\n",
    "                non_overlap_genes.append(feature['name'])\n",
    "    \n",
    "    non_overlap_df = features_df.query('name in @non_overlap_genes').reset_index(drop = True)\n",
    "\n",
    "    #create temp gene fasta\n",
    "    with open('temp/{}_nonoverlap.fasta'.format(species_name), 'w') as f:\n",
    "        for i, gene in non_overlap_df.iterrows():\n",
    "            name = gene['name']\n",
    "            sequence = gene['translation']\n",
    "            f.write(f'>{name}\\n{sequence}\\n')\n",
    "    \n",
    "    #run mmseqs from temp fasta files\n",
    "    client = mmseqs.MMSeqs()\n",
    "    client.databases.create(species_name, species_name, 'temp/{}_nonoverlap.fasta'.format(species_name))\n",
    "    \n",
    "    #query this database with housekeeping genes\n",
    "    results = client.databases[0].search_file(\"temp/housekeeping.fasta\", search_type=\"proteins\", headers = ['target_sequence_id','e_value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "1d595785",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2300"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
